src/stm.ts

```diff
@@ -0,0 +1,515 @@
+  }
+
+  /**
+   * Inserts a delete operation.
+   */
+  public addDeletion(req: RPC.IDeleteRangeRequest) {
+    if (req.range_end) {
+      this.ops.push({ req, op: WriteKind.DeleteRange, range: new Range(req.key!, req.range_end) });
+    } else {
+      this.ops.push({ req, op: WriteKind.DeleteKey, key: req.key! });
+    }
+  }
+}
+
+/**
+ * BasicTransaction is the base wrapper class for a transaction. It implements
+ * the necessary mechanics for Repeatablereads
+ * and ReadCommitted isolation levels.
+ */
+class BasicTransaction {
+  public readonly writeSet = new WriteSet();
+  public readonly readSet = new ReadSet();
+
+  constructor(protected readonly options: ISTMOptions) {}
+
+  /**
+   * Gets the range in a transaction-y way!
+   */
+  public range(kv: RPC.KVClient, req: RPC.IRangeRequest): Promise<RPC.IRangeResponse> {
+    this.assertReadInvariants(req);
+
+    const existingWrite = this.writeSet.findExistingWrite(req.key!);
+    if (existingWrite !== null) {
+      return Promise.resolve(existingWrite);
+    }
+
+    req.serializable = true;
+    return this.readSet.runRequest(kv, req);
+  }
+
+  /**
+   * Schedules the put request in the writeSet.
+   */
+  public put(req: RPC.IPutRequest): Promise<RPC.IPutResponse> {
+    this.assertNoOption('put', req, 'lease', 'prev_kv');
+    this.writeSet.addPut(req);
+    return Promise.resolve(<any>{});
+  }
+
+  /**
+   * Schedules the put request in the writeSet.
+   */
+  public deleteRange(req: RPC.IDeleteRangeRequest): Promise<RPC.IDeleteRangeResponse> {
+    this.assertNoOption('delete', req, 'prev_kv');
+    this.writeSet.addDeletion(req);
+    return Promise.resolve({
+      header: <any>undefined,
+      deleted: '1',
+      prev_kvs: [],
+    });
+  }
+
+  protected assertReadInvariants(range: RPC.IRangeRequest) {
+    this.assertNoOption(
+      'read',
+      range,
+      'revision',
+      'min_mod_revision',
+      'max_mod_revision',
+      'min_create_revision',
+      'max_create_revision',
+    );
+  }
+
+  protected assertNoOption<T>(req: string, obj: T, ...keys: (keyof T)[]) {
+    keys.forEach(key => {
+      if (obj[key] !== undefined) {
+        throw new Error(`"${key}" is not supported in ${req} requests within STM transactions`);
+      }
+    });
+  }
+}
+
+/**
+ * BasicTransaction is the class for serializable transactions. It implements
+ * the necessary mechanics for SerializableSnapshot
+ * and Serializable isolation levels.
+ */
+class SerializableTransaction extends BasicTransaction {
+  private firstRead: Promise<RPC.IRangeResponse> | null;
+
+  constructor(options: ISTMOptions, kv: RPC.KVClient) {
+    super(options);
+    options.prefetch.forEach(key => {
+      this.range(kv, { key: toBuffer(key) }).catch(() => undefined);
+    });
+  }
+
+  /**
+   * @override
+   */
+  public range(kv: RPC.KVClient, req: RPC.IRangeRequest): Promise<RPC.IRangeResponse> {
+    this.assertReadInvariants(req);
+
+    this.assertNoOption(
```

Comment:
should there be a check for `range_end` here or is it already filtered out?

Reply:
Sure. We only actually expose the SingleRangeBuilder from the transactor but it's possible someone could hack it in there, thinking it should.